File: Exercism/java/triangle/src/main/java/Triangle.java
/**
 * Created by cajetan on 5/16/17.
 */
public class Triangle {
    private double sideA;
    private double sideB;
    private double sideC;

    public Triangle(double sideA, double sideB, double sideC) throws TriangleException {
        this.sideA = sideA;
        this.sideB = sideB;
        this.sideC = sideC;

        if (sideA <= 0 || sideB <= 0 || sideC <= 0 ||
                sideA+sideB <= sideC || sideB+sideC <= sideA ||
                sideC+sideA <= sideB) {
            throw new TriangleException();
        }
    }

    public TriangleKind getKind() {
        if (sideA == sideB && sideB == sideC)
            return TriangleKind.EQUILATERAL;

        if (sideA == sideB || sideB == sideC || sideA == sideC)
            return TriangleKind.ISOSCELES;

        return TriangleKind.SCALENE;
    }
}

File: Exercism/go/triangle/triangle.go
// This is a "stub" file.  It's a little start on your solution.
// It's not a complete solution though; you have to write some code.

// Package triangle should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package triangle

import (
	"math"
)

type Kind string

const (
	NaT = "NaT" // not a triangle
	Equ = "Equ" // equilateral
	Iso = "Iso" // isosceles
	Sca = "Sca" // scalene
)

// KindFromSides should have a comment documenting it.
func KindFromSides(a, b, c float64) Kind {
	if math.IsInf(a, 0) || math.IsInf(b, 0) || math.IsInf(c, 0) {
		return NaT
	}

	if math.IsNaN(a) || math.IsNaN(b) || math.IsNaN(c) {
		return NaT
	}

	if a <= 0 || b <= 0 || c <= 0 || a+b < c || b+c < a || c+a < b {
		return NaT
	}

	if a == b && b == c {
		return Equ
	}

	if a == b || b == c || a == c {
		return Iso
	}

	if a != b && b != c && a != c {
		return Sca
	}

	return NaT
}
